algorithms/src/snark/varuna/data_structures/circuit_proving_key.rs

```diff
@@ -58,6 +58,12 @@ impl<E: PairingEngine, SM: SNARKMode> FromBytes for CircuitProvingKey<E, SM> {
     }
 }
 
+impl<E: PairingEngine, SM: SNARKMode> PartialEq for CircuitProvingKey<E, SM> {
+    fn eq(&self, other: &Self) -> bool {
+        self.circuit.id == other.circuit.id
```

Comment:
I'm not convinced this is safe.

There is no check in the `FromBytes` impl that this is computed correctly, and corresponds to the objects involved.

I would suggest auto-deriving `PartialEq, Eq` on the entire struct.

Reply:
This would be very expensive. In the `FromBytes` impl I believe we reconstruct the `id` by hashing.

Reply:
> I'm not convinced this is safe.
Not sure what the attack vector is. I'm not aware of scenario's where a user should accept a CircuitProvingKey from a third party.

But even if they did, as Pratyush mentions, the user computes the `circuit_id` themselves which the user and verifiers should always check against. In theory the third party could put bogus precomputed values (index polys, (i)fft_precomputation) in the proving key to make the proof fail to be generated - not to accidentally prove a different statement.